client/clientservice/src/state_snapshot_service.cpp

```diff
@@ -22,27 +28,388 @@ using vmware::concord::client::statesnapshot::v1::StreamSnapshotResponse;
+    std::chrono::milliseconds curr_timeout{time_left};
+    if (curr_timeout.count() > 0) {
+      timeout = curr_timeout;
+    }
+  }
+
+  WriteConfig write_config{RequestConfig{false, 0, 1024 * 1024, timeout, "snapshotreq", "", false, true},
+                           bft::client::LinearizableQuorum{}};
+  StateSnapshotResponse snapshot_response;
+  grpc::Status return_status;
+  auto callback =
+      [&snapshot_response, &return_status, &write_config](concord::client::concordclient::SendResult&& send_result) {
+        ResponseType<StateSnapshotResponse> reponse;
+        getResponseSetStatus(std::move(send_result),
+                             reponse,
+                             return_status,
+                             write_config.request.correlation_id,
+                             false,
+                             "concord.client.clientservice.state_snapshot_service.GetRecentSnapshot.callback");
+        snapshot_response = std::get<StateSnapshotResponse>(reponse.response);
+      };
+
+  ReconfigurationRequest rreq;
+  StateSnapshotRequest cmd;
+  cmd.checkpoint_kv_count = 0;
+  cmd.participant_id = client_->getSubscriptionId();
+  rreq.command = cmd;
+  bft::client::Msg message;
+  concord::messages::serialize(message, rreq);
+  client_->send(write_config, std::move(message), callback);
+  if (return_status.ok() && snapshot_response.data.has_value()) {
+    response->set_snapshot_id(snapshot_response.data->snapshot_id);
+    response->set_event_group_id(snapshot_response.data->event_group_id);
+    response->set_key_value_count_estimate(snapshot_response.data->key_value_count_estimate);
+    auto* ledger_time = response->mutable_ledger_time();
+    *ledger_time = google::protobuf::util::TimeUtil::GetEpoch();
+  }
+  return return_status;
 }
 
 Status StateSnapshotServiceImpl::StreamSnapshot(ServerContext* context,
                                                 const StreamSnapshotRequest* proto_request,
                                                 ServerWriter<StreamSnapshotResponse>* stream) {
-  // TODO: Add implementation
-  return grpc::Status(grpc::StatusCode::UNIMPLEMENTED, "StreamSnapshot");
+  ConcordAssertNE(proto_request, nullptr);
+  concord::client::concordclient::StateSnapshotRequest request;
+  request.snapshot_id = proto_request->snapshot_id();
+  if (proto_request->has_last_received_key()) {
+    request.last_received_key = proto_request->last_received_key();
+  }
+  std::shared_ptr<SnapshotQueue> update_queue = std::make_shared<BasicSnapshotQueue>();
+
+  client_->getSnapshot(request, update_queue);
+
+  Status status = grpc::Status(grpc::StatusCode::UNAVAILABLE, "Service not available");
+  bool is_end_of_stream = false;
+  auto accumulated_hash = singleHash(std::string{});
+  while (!context->IsCancelled()) {
+    StreamSnapshotResponse response;
+    std::unique_ptr<SnapshotKVPair> update;
+    try {
+      update = update_queue->pop();
```

Comment:
This will block until the timeout, since the GrpcConnection object times out and gives a timeout result. Trypop is better for updates and pop is better for stream reads.

Reply:
That is an implementation detail and an assumption you are putting into this line of code. The code can block forever. The assumption is that if the underlying gRPC connection to the replica times out, you will push a timeout update/exception. However, what if the underlying code changes? What if the gRPC connection times out but we forget to populate that information? Or, ...
Bottom line, this code can block forever because there is no timeout specified. In addition, what if the caller decides to cancel the gRPC stream? This will never be checked unless we get an update/exception.

Reply:
ok I will change this to tryPop(). Thanks @robem